### PR TITLE
Improve package name completion (fix #2456)

### LINKED ISSUE #2456: Suggestion for package title adds a `__test` instead of `_test` only
- Plugin version (or commit hash): 0.11.1288
IntelliJ IDEA (Minerva) IU-145.844.1
Build #IU-145.844, built on April 5, 2016
JRE: 1.8.0_76-release-b18 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o

- What are you trying to do?
Have a package named `package_test`
- What would you expect to happen?
The only suggestion to be `package_test`
- What happens?
I have two suggestions presented: `package__test` and `package_test` (first one is wrong).

snapshot25

### LINKED ISSUE #2458: Completion suggests items that are not present in the current context for structs
- Plugin version (or commit hash): 0.11.1288
IntelliJ IDEA (Minerva) IU-145.844.1
Build #IU-145.844, built on April 5, 2016
JRE: 1.8.0_76-release-b18 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o

- What are you trying to do?
Write a struct initialization

- What would you expect to happen?
Have the struct fields appear in the list

- What happens?
Values that are not related to the struct fields appear in the list.

```go
package main

type demo struct{
	FieldA string
	Formatter string
	Hello string
}

func main() {
	a := demo{
		<caret>
	}
}
```

Normal autocompletion while typing
snapshot86

CTRL+SPACE (smart completion?)
snapshot87

### PR DESCRIPTION
**Pull request by dlsniper**

### CODE REVIEW

**Comment by zolotov** [comment]
Sorry, the PR is unreadable, nested loops, strange string length comparations, a lot of undescribed if-branches, strange startWith's, etc. I cannot understand it, which means that I won't be able to support it in future

File: tests/com/goide/completion/GoCompletionTest.java [hunk 1]
```diff
@@ -522,15 +522,63 @@ public void testPackageNames() {
...
+  @SuppressWarnings("ConstantConditions")
+  public void testPackageNamesFromMainPackagePartialName() {
+    myFixture.configureByText("foo.go", "package main");
+    myFixture.configureByText("foo_test.go", "package m<caret>");
+    myFixture.completeBasic();
+    assertSameElements(myFixture.getLookupElementStrings(), "main", "main_test");
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  public void testPackageNamesFromMainPackagePartialName2() {
```
Comment (zolotov) [inline tests/com/goide/completion/GoCompletionTest.java]: `testPackageNamesFromMainPackagePartialName` is the same, no?

Comment (dlsniper) [inline tests/com/goide/completion/GoCompletionTest.java]: Yes, I've missed that one.

**Comment by dlsniper** [comment]
> Sorry, the PR is unreadable, nested loops, strange string length comparations, a lot of undescribed if-branches, strange startWith's, etc. I cannot understand it, which means that I won't be able to support it in future

I agree it should definitely be improved.
What I've tried here was to get the tests for the issue in place and see how we can have a fix for it.
I've been thinking on the algorithm behind this and I think it should be like this (pseudo-code wise):

```
currentPackage = file.getCurrentPackage()
packagesInDirectory = getPackagesInDirectory()

packages = []
for package : packagesInDirectory {
  if package.equals(currentPackage) {
    if !isTestFile && file.getDirectory().Files().size() > 1 {
      packages[] = package
    } else if !isTestFile {
      packages[] = package
    }
  } else if package.startsWith(currentPackage) {
    packages[] =  packages
  }
  else if !isTestFile && currentPackage.startsWith(package) {
    // should we suggest this?
  }
}

if (packages.size < 1) {
  packages[] = inferPackageFromFileDirName()
  packages[] =  "main"
}

for package : packages {
  completion.add(package)
  if (isTestFile) completion.add(package+"_test")
}
```

I will need to run this for the existing scenarios and see if the algorithm works and covers everything.
The most complex part is trying to separate the package that's just defined in the current file (since the IDE will recognize the package before the completion is called).

**Comment by zolotov** [comment]
> if !isTestFile && file.getDirectory().Files().size() > 1 {
      packages[] = package
    } else if !isTestFile {
      packages[] = package
    }

This actually means `if (!isTestsFile) packages[] = package`

Anyway, pseudo-code looks much better. And how about just replacing this code:
```
if (!packageName.equals(currentPackageName)) {
  result.addElement(packageLookup(packageName, GoCompletionUtil.PACKAGE_PRIORITY - 1));
}
if (isTestFile) {
  result.addElement(packageLookup(packageName + GoConstants.TEST_SUFFIX, GoCompletionUtil.PACKAGE_PRIORITY));
}
```

to

```
if (!packageName.equals(currentPackageName)) {
  result.addElement(packageLookup(packageName, GoCompletionUtil.PACKAGE_PRIORITY - 1));
  if (isTestFile) {
    result.addElement(packageLookup(packageName + GoConstants.TEST_SUFFIX, GoCompletionUtil.PACKAGE_PRIORITY));
  }
}
```

**Pull request closed without merging**